tools/Debugger/network-debugger.cpp

```diff
@@ -0,0 +1,242 @@
+                  llvm::cl::Optional, llvm::cl::cat(debuggerTestCat));
+
+llvm::cl::opt<bool> fuseScaleOffsetFp16Opt(
+    "glow_global_fused_scale_offset_fp16",
+    llvm::cl::desc(
+        "Enable fp16 lowering for all op inputs using fused scale/offset"),
+    llvm::cl::Optional, llvm::cl::cat(debuggerTestCat));
+
+llvm::cl::opt<bool>
+    ClipFp16Opt("glow_clip_fp16",
+                llvm::cl::desc("Force glow to clip fp16 values to min/max"),
+                llvm::cl::Optional, llvm::cl::cat(debuggerTestCat));
+
+llvm::cl::opt<bool>
+    forceFP16AccumSLSOpt("glow_global_force_sls_fp16_accum",
+                         llvm::cl::desc("Force FP16 accumulation for SLS ops"),
+                         llvm::cl::Optional, llvm::cl::cat(debuggerTestCat));
+
+llvm::cl::opt<bool> enablePartialTensor("glow_enable_partial_tensor",
+                                        llvm::cl::desc("Enable partial tensor"),
+                                        llvm::cl::Optional,
+                                        llvm::cl::cat(debuggerTestCat));
+#define DEBUG_TYPE "verifier"
+
+void parseCommandLine(int argc, char **argv) {
+  llvm::sys::PrintStackTraceOnErrorSignal(argv[0]);
+  llvm::cl::ParseCommandLineOptions(
+      argc, argv,
+      " Network Debugger tool, part of the Glow compiler\n\n"
+      "Glow is a compiler for neural network accelerators.\n");
+}
+
+::ONNX_NAMESPACE::GraphProto parseIO(const std::string &filename) {
+  ::ONNX_NAMESPACE::GraphProto g;
+  std::ifstream ff(filename, std::ios::in | std::ios::binary);
+  CHECK(ff) << "Can't find the input file for " << filename.c_str();
+  google::protobuf::io::IstreamInputStream fileStream(&ff);
+  google::protobuf::io::CodedInputStream codedStream(&fileStream);
+  codedStream.SetTotalBytesLimit(MAX_PROTO_SIZE, MAX_PROTO_SIZE);
+  bool yes = g.ParseFromCodedStream(&codedStream);
+  CHECK(yes) << "Failed to parse GraphProto";
+  return g;
+}
+
+// This holds the tensor that actually owns the data for all the partial
+// inputs.
+std::vector<Tensor> partialTensorPayloads;
+
+// Utility functions:
+void convertNetwork(Function *F, bool verbose = true) {
+  CompilationContext cctx;
+  PrecisionConfiguration precConfig;
+  if (globalFp16Opt) {
+    precConfig.convertToFP16 = globalFp16Opt;
+    if (verbose)
+      llvm::outs() << "Conversion to fp16 enabled\n";
+  }
+  if (fuseScaleOffsetFp16Opt) {
+    precConfig.convertFusedToFP16 = fuseScaleOffsetFp16Opt;
+    if (verbose)
+      llvm::outs() << "Conversion of fused scales/offsets to fp16 enabled\n";
+  }
+  if (ClipFp16Opt) {
+    precConfig.clipFP16 = ClipFp16Opt;
+    if (verbose)
+      llvm::outs() << "Clipping to fp16 enabled\n";
+  }
+  if (forceFP16AccumSLSOpt) {
+    precConfig.forceFP16AccumSLS = true;
+    if (verbose)
+      llvm::outs() << "Forcing fp16 accumulation for SLS ops enabled\n";
+  }
+  // Convert Network to fp16.
+  glow::convertFunctionToFloat16(F, precConfig);
```

Comment:
The idea here was to not compile the module, just optimize it to remove convertTo nodes that get add after we convert it to fp16.


> - are they going to be dumped from Glow? Are they going to have gone through Glow's optimization pipeline already?
If I understood your question correctly, no. they will just be loaded from the protobuf file. The optimizer is just to get rid of extra nodes. I didn't want the module ot be compiled because the  Comparator classes ar going to further modify it (by inserting hooks) then compile. You can't compile a module twice